### ISSUE TITLE
Rewrite BZDB to be a smarter structure

### ISSUE THREAD

**Issue by allejo** [BZFlag-Dev/bzflag#248]
Our `StateDatabase` needs a lot of rethinking. Right now, calling a simple `bz_getBZDBInt` calls `BZDB.eval` which is this behemoth:

https://github.com/BZFlag-Dev/bzflag/blob/040b45d94a07ccbb51b0434502f964886fede539/src/common/StateDatabase.cxx#L412-L450

I'd like to suggest that this be rewritten to use smarter data structures; e.g. a wrapper around separate `std::map`s for each data type (int, double, bool, string).

Thoughts for any other implementations instead of having to "evaluate" all of these values?

**Comment by blast007** [comment]
Yeah, currently we have the parsing overhead every time we want to get a value. I'd have to do some code analysis but I would think we generally only reference a variable as a single type, so having all the data types for every variable may be overkill. I envision more of a full blown BZDBCache that covers every variable.

Since, as you mention, it's possible for a variable to reference another variable, we could have a list of references to quickly know which other variables we need to update, or maybe more simply just a list of variables that *do* reference other variables.

For plugins I'd have to think about it more. Maybe plugins could register custom variables with a callback.

It would also be worth thinking if we want to keep using BZDB for local client configuration or if we'd want to move to a configuration class (which would help with splitting the configuration into separate files).

**Comment by Zehra** [comment]
The variables themselves remain mostly unchanging, therefore checks should be made to see if a variable has changed to minimize overhead of evaluating variables.

This could be done within a simple lookup array.
```
    if (isChanged[bzdbVariable] == true) {
        evaluateBZDB(bzdbVariable);
        isChanged[bzdbVariable] = false;
    }
    return value[bzdbVariable];
```
However it would be perhaps a bit tricky to implement this for bzdb variables which use/reference other bzdb variables in comparisons.

Perhaps something similar to the following may work?
```
if (bzdbExists(bzdbVariable)) {
    if (isChanged[bzdbVariable] == true) {
        evaluateBZDB(bzdbVariable);
        isChanged[bzdbVariable] = false;
    }
    return value[bzdbVariable];
}
```

As for determing the type of value, the following may work:
```
variableType = bzdbType[bzdbVariable];
switch(variableType){
    case int:{}break;
    case long:{}break;
    case float:{}break;
    case double:{}break;
    case bool:{state}break;
    default:{}break;
}

```

The implementation within the switch statement is left as-is, pending further data.
Although one possible solution would be the following:
```
variableType = bzdbType[bzdbVariable];
slotInSpecifiedArray = bzdbSlotInArray[bzdbVariable];
switch(variableType){
    case int:{ return intArray[slotInSpecifiedArray];}break;
    case long:{ return longArray[slotInSpecifiedArray];}break;
    case float:{ return floatArray[slotInSpecifiedArray];}break;
    case double:{ return doubleArray[slotInSpecifiedArray];}break;
    case bool:{ return boolArray[slotInSpecifiedArray];}break;
    default:{}break;
}

```

Although the whole thing could be most easily optimized by simply making a default BZDB and custom BZDB, of which only custom is enabled and used whenever there is a custom BZDB variable registered.

```
if (bzdbVariable exists in defaultBZDB) {
    return data;
} else {
    if (EnabledCustomBZDB) {
        if (bzdbVariable exists in CustomBZDB) {
            return data;
        }
    }
    return does not exist;
}
```

**Comment by jwmelto** [comment]
@Zehra that smells of premature optimization. Is our goal here performance or simplicity/readability?

The basic challenge is that C++ doesn't have heterogeneous containers. A fairly straightforward alternative is something like [jsoncpp](https://github.com/open-source-parsers/jsoncpp).

**Comment by Zehra** [comment]
@jwmelto I'm not entirely sure, but it seems it should be both, with more optimized logistics leading to better readability.

Although my previous comment makes a few assumptions and assumes the following:
1. Much of the overhead is caused by evaluation of values in a BZDB variable.
2. The variables themselves are not frequently added/removed.
3. Looking up an item in an array is more efficient and uses less resources than evaluation of a variable.
4. The value a BZDB variable has does not tend to change quite often, if at all.

* **BZDB variable** is used in reference to items such as `_worldSize`, `_tankSpeed`.
* **value* is used in reference to a BZDB variable's type and the value held within it.
(a.k.a. *`_worldSize` at `800.0`* with `800.0` being the **value** of the **BZDB variable** `_worldSize`.)

Inaccuracies in my previous comment:
* "The variables themselves remain mostly unchanging,"
(Should be substituted with: "The BZDB variable values themselves remain mostly unchanging,")

The above is solely an optimization with logistics. If any of the my assumptions are incorrect, my previous comment is invalid and incorrectly calculates/factors things.

Although, if JSON can be used, I'm not certain, but it seems that SQLite may be better suited, although I haven't looked that much into it.

**Comment by jwmelto** [comment]
In the pre-standard days of C++, I picked up a statement from the newsgroups. I can't find a reference to it anymore, but the way I internalized it was this:
> The Law of Premature Optimization states that "Every programmer intuitively knows where the bottlenecks in his program are... and he is always wrong."

The point, which is borne out by Knuth's seminal quote:
> There is no doubt that the grail of efficiency leads to abuse. Programmers waste enormous amounts of time thinking about, or worrying about, the speed of noncritical parts of their programs, and these attempts at efficiency actually have a strong negative impact when debugging and maintenance are considered. We should forget about small efficiencies, say about 97% of the time: premature optimization is the root of all evil.
Yet we should not pass up our opportunities in that critical 3%. A good programmer will not be lulled into complacency by such reasoning, he will be wise to look carefully at the critical code; but only after that code has been identified. 

is that the only inefficiency worth spending time optimizing is that which has been proven to be worth the effort.

Of course, on a FOSS project, people are free to spend their energy where they wish. I'm just not sure we've proven that BDZB is a demonstrable performance issue.

There is a philosophical design choice underlying BZDB; namely that runtime evaluation is preferable to compile time. If I were designing from scratch, I'd have a structure with all the possible values defined in the appropriate types. The population/modification of said structure is an interface issue, and could be via file/database, API, command-line, or smoke signal.

I'm sure there are reasons that I'm not sensitive to that make the current lazy evaluation highly desirable, but then should we really gnash our teeth about performance or the ugliness of the API? YMMV

**Comment by macsforme** [comment]
I envision a BZDB where variables are stored in their native type and not in a map (similar to the current BZDB cache, as @blast007 mentioned).  This ought to be both simpler, *and* more efficient.  As others mentioned, there are some edge cases to be addressed, such as variables that reference other variables, but ultimately I believe that a much simpler class will fit our needs better.
